Implement permission request for Apple embedded platforms, fix microphone input

migueldeicaza opened the pull request:
Fixes https://github.com/godotengine/godot-proposals/issues/12563 but also fixes the "Recording audio on iOS does not work" from 33885.

This now addresses all issues I spotted in https://github.com/godotengine/godot/issues/33885 and I have tested it against the Godot recording sample and it finally works on iOS.

migueldeicaza committed:
Implement permission request for Apple embedded platforms, fixes https://github.com/godotengine/godot/issues/107229

syntaxerror247 commented on doc/classes/OS.xml:
```diff
@@ -781,8 +781,9 @@
+				- [code]OS.request_permission("appleembedded.permission.AUDIO_RECORD")[/code]
 				[b]Note:[/b] On Android, permission must be checked during export.
-				[b]Note:[/b] This method is implemented on Android and macOS.
+				[b]Note:[/b] This method is implemented on Android, macOS and Apple embedded platforms.
```
I think, this should mention iOS and visionOS. Term `Apple embedded platforms` seems to be only used internally.
```suggestion
				[b]Note:[/b] This method is implemented on Android, iOS, macOS, and visionOS.
```

migueldeicaza committed:
Instead of allocating and guessing the size of the number of frames
that the OS will provide us, cope with the fact that iOS will adjust
the number of frames dynamically based on your session type.

This should fix #33885

migueldeicaza committed:
Address feedback

migueldeicaza commented:
This pull request now also includes a fix for 33885

Calinou commented:
I wonder how https://github.com/godotengine/godot/pull/106305 has been confirmed to be working by its author, since I expect camera access to also require a permission just like microphone access.

shiena commented:
> I wonder how #106305 has been confirmed to be working by its author, since I expect camera access to also require a permission just like microphone access.

I've created a test application where I've incorporated CameraTexture into the scene tree. As a result, you'll notice a camera access permission dialog appearing when the app launches.

bruvzg reviewed (approved):
Tested on 10. gen. iPad, recording seems to be working (note for testing, appropriate category should set in `audio/general/ios/session_category` project setting and permission description added in export config for it to work).
